### ISSUE TITLE
Support structured logging

### ISSUE THREAD

**Issue by mlipovsky** [nats-io/nats-server#1240]
- [x] Feature Request
 
## Feature Requests
Hi,
we would like to switch the logging format to structured (json). For easier analysis of logs.
 
If you would welcome this feature, we could do it and create a pull request.

#### Use Case:
Projects with automated log analysis.
 
#### Proposed Change:
Add a new configuration to create a structured logger.

**Comment by derekcollison** [comment]
Which log statements in particular? Of course there is a large ecosystem around NATS and tooling that is built around the old format.

However, as we move more towards black box operations (debugging things without access to the server or logs themselves) we are adding new things that can take advantage of JSON structuring and delivery.

**Comment by mcuadros** [comment]
It Will be great if can be added a flag to turn the logging into JSON, otherwise, platforms like Google Logging, can't set the severity of the messages properly.

**Comment by derekcollison** [comment]
Can you point us to a link on the Google logging you are using?

**Comment by mcuadros** [comment]
I am working on this problem right now, in GKE if you don't use the custom format of [Stackdrive/Google Logging](https://cloud.google.com/logging/docs) you need to deploy a [custom configuration](https://cloud.google.com/solutions/customizing-stackdriver-logs-fluentd), ending in another component to maintain beside that you need to do a regexp to parse logs.

**Comment by derekcollison** [comment]
Looping in @wallyqs and @ripienaar

**Comment by ripienaar** [comment]
I'm a big fan of structured logging, the problem is this is a deep pit as every platform wants their own crazy formats

**Comment by mcuadros** [comment]
But at least if it is structured, you don't need to create a regexp. Also, you can use logrus, where almost every single format it's already implemented, and you can implement the format for the bigger clouds, with a small effort.

**Comment by ripienaar** [comment]
Yeah, in my own projects I use logrus and it's been a life saver for this reason. Pretty weary of new dependencies here though is the problem :)

**Comment by mcuadros** [comment]
yep, that's why I suggested a generic json.

**Comment by jcogilvie** [comment]
Several people have offered to implement it.  Perhaps one of them can?  (I can speak highly of zap as a solution.)  

My use case is that NATS is included as a helm dependency in something else I'm using, and now it's polluting my otherwise nicely structured logs in datadog.  And while yes, I can write a custom parse rule, it's significantly easier to just ingest a structured object so that everything isn't ingested as an error.

**Comment by NickLarsenNZ** [comment]
I would like to be able to use _logfmt_ instead of _JSON_, but regardless this is really the job of the logger implementation.

It might be worth considering [OpenTelemetry](https://opentelemetry.io/docs/instrumentation/go/) (not only for logging, but also traces and metrics).
It looks like logs aren't supported yet for Go, but that's the same case for Rust. (I've found it helpful using the tracing crate for logging with attributes in libs, while then using a tracing subscriber to deal with how logs are exposed along with attributes in the resulting app - I imagine it to be relatively similar for the Go SDK).

**Comment by Kaarel** [comment]
+1 OpenTelemetry
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md

**Comment by yordis** [comment]
About the topic, and keeping OTEL in mind,

Could most things be traces rather than logs? You can compose a log out of a trace, but it is tricky to do the other way around!

So maybe what most people want is tracing.

**Comment by ktarplee** [comment]
It seems to me that the way to implement this now is with the standard library package [log/slog](https://pkg.go.dev/log/slog).
